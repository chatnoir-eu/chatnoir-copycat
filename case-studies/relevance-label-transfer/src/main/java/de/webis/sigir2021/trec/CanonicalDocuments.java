package de.webis.sigir2021.trec;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.io.IOUtils;
import org.json.JSONArray;
import org.json.JSONObject;

import lombok.SneakyThrows;

public class CanonicalDocuments {
	
	private static Map<String, String> DOC_TO_CANONICAL = null;
	
	public static synchronized String canonicalDocument(String id) {
		if (DOC_TO_CANONICAL == null) {
			DOC_TO_CANONICAL = buildDocToCanonical();
		}
		
		return DOC_TO_CANONICAL.getOrDefault(id, id);
	}

	private static Map<String, String> buildDocToCanonical() {
		Map<String, String> ret = new LinkedHashMap<>();
		ret.putAll(buildDocToCanonical("/clueweb09-web-track-content-equivalent-groups.jsonl"));
		ret.putAll(buildDocToCanonical("/clueweb12-web-track-content-equivalent-groups.jsonl"));
		
		return ret;
	}
	
	@SneakyThrows
	private static Map<String, String> buildDocToCanonical(String resource) {
		List<String> jsonLines = IOUtils.readLines(CanonicalDocuments.class.getResourceAsStream(resource));
		Map<String, String> ret = new LinkedHashMap<>();
		
		for(String jsonLine: jsonLines) {
			List<String> equivalentDocs = equivalentDocs(jsonLine);
			
			for(int i=1; i<equivalentDocs.size(); i++) {
				ret.put(equivalentDocs.get(i), equivalentDocs.get(0));
			}
		}
		
		return ret;
	}

	private static List<String> equivalentDocs(String json) {
		List<String> ret = new ArrayList<>();
		JSONArray ids = new JSONObject(json).getJSONArray("ids");
		
		for(int i=0; i< ids.length(); i++) {
			ret.add(ids.getString(i));
		}
		
		Collections.sort(ret);
		
		return ret;
	}

	public static boolean isDuplicate(String id) {
		return id == null || !id.equals(canonicalDocument(id));
	}
}
