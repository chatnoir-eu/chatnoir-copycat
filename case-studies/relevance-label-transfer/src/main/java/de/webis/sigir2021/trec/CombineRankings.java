package de.webis.sigir2021.trec;

import static io.anserini.index.generator.LuceneDocumentGenerator.FIELD_ID;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.lucene.document.Document;
import org.apache.lucene.document.Field.Store;
import org.apache.lucene.document.StringField;

import de.webis.trec_ndd.spark.RunLine;
import de.webis.trec_ndd.trec_collections.QrelEqualWithoutScore;
import de.webis.trec_ndd.trec_collections.SharedTask;
import io.anserini.index.generator.LuceneDocumentGenerator;
import io.anserini.rerank.ScoredDocuments;
import io.anserini.rerank.lib.ScoreTiesAdjusterReranker;
import lombok.Data;
import lombok.SneakyThrows;

@Data
public class CombineRankings {
	private static final List<Float> WEIGHTS = Arrays.asList(0.0f, 0.2f, 0.4f, 0.6f, 0.8f, 1.0f);
	
	public static void main(String[] args) {
		Path baseDir = Paths.get("/mnt/ceph/storage/data-in-progress/kibi9872/sigir2021/data-13-10-2020/rankings/");
		
		createAllRuns(
			baseDir.resolve("topics-1-50-cw09"),
			"webis_warc_clueweb09_003",
			TransferredSharedTasks.WEB_2009_DUPLICATE_FREE,
			baseDir.resolve("topics-1-50-cw09").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-51-100-cw09"),
			"webis_warc_clueweb09_003",
			TransferredSharedTasks.WEB_2010_DUPLICATE_FREE,
			baseDir.resolve("topics-51-100-cw09").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-101-150-cw09"),
			"webis_warc_clueweb09_003",
			TransferredSharedTasks.WEB_2011_DUPLICATE_FREE,
			baseDir.resolve("topics-101-150-cw09").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-151-200-cw09"),
			"webis_warc_clueweb09_003",
			TransferredSharedTasks.WEB_2012_DUPLICATE_FREE,
			baseDir.resolve("topics-151-200-cw09").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-1-50-cw12"),
			"webis_warc_clueweb12_011",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2009_TO_CW_2012,
			baseDir.resolve("topics-1-50-cw12").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-51-100-cw12"),
			"webis_warc_clueweb12_011",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2010_TO_CW_2012,
			baseDir.resolve("topics-51-100-cw12").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-101-150-cw12"),
			"webis_warc_clueweb12_011",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2011_TO_CW_2012,
			baseDir.resolve("topics-101-150-cw12").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-151-200-cw12"),
			"webis_warc_clueweb12_011",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2012_TO_CW_2012,
			baseDir.resolve("topics-151-200-cw12").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-1-50-cw12wb12"),
			"clueweb12-and-wayback12",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2009_TO_CW_WB_2012,
			baseDir.resolve("topics-1-50-cw12wb12").resolve("final-rankings")
		);
			
		createAllRuns(
			baseDir.resolve("topics-51-100-cw12wb12"),
			"clueweb12-and-wayback12",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2010_TO_CW_WB_2012,
			baseDir.resolve("topics-51-100-cw12wb12").resolve("final-rankings")
		);
			
		createAllRuns(
			baseDir.resolve("topics-101-150-cw12wb12"),
			"clueweb12-and-wayback12",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2011_TO_CW_WB_2012,
			baseDir.resolve("topics-101-150-cw12wb12").resolve("final-rankings")
		);
			
		createAllRuns(
			baseDir.resolve("topics-151-200-cw12wb12"),
			"clueweb12-and-wayback12",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2012_TO_CW_WB_2012,
			baseDir.resolve("topics-151-200-cw12wb12").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-201-250-cw12"),
			"webis_warc_clueweb12_011",
			TransferredSharedTasks.WEB_2013_DUPLICATE_FREE,
			baseDir.resolve("topics-201-250-cw12").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-251-300-cw12"),
			"webis_warc_clueweb12_011",
			TransferredSharedTasks.WEB_2014_DUPLICATE_FREE,
			baseDir.resolve("topics-251-300-cw12").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-1-50-cc15"),
			"webis_warc_commoncrawl15_002",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2009_TO_CC_2015,
			baseDir.resolve("topics-1-50-cc15").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-51-100-cc15"),
			"webis_warc_commoncrawl15_002",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2010_TO_CC_2015,
			baseDir.resolve("topics-51-100-cc15").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-101-150-cc15"),
			"webis_warc_commoncrawl15_002",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2011_TO_CC_2015,
			baseDir.resolve("topics-101-150-cc15").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-151-200-cc15"),
			"webis_warc_commoncrawl15_002",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2012_TO_CC_2015,
			baseDir.resolve("topics-151-200-cc15").resolve("final-rankings")
		);
		
		createAllRuns(
			baseDir.resolve("topics-201-250-cc15"),
			"webis_warc_commoncrawl15_002",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2013_TO_CC_2015,
			baseDir.resolve("topics-201-250-cc15").resolve("final-rankings")
		);

		createAllRuns(
			baseDir.resolve("topics-251-300-cc15"),
			"webis_warc_commoncrawl15_002",
			TransferredSharedTasks.TRANSFER_FROM_WEB_2014_TO_CC_2015,
			baseDir.resolve("topics-251-300-cc15").resolve("final-rankings")
		);
	}
	
	@SneakyThrows
	private static void createAllRuns(Path baseRunDir, String prefix, SharedTask task, Path outputDir) {
		CombineRankings combiner = new CombineRankings(baseRunDir.toFile(), prefix, task);
		
		for(float bodyWeight: WEIGHTS) {
			for(float titleWeight: WEIGHTS) {
				for(float metaWeight: WEIGHTS) {
					String output = combiner.combine(bodyWeight, titleWeight, metaWeight);
					Path outputFile = outputDir.resolve(CombineRankings.runTag(prefix, bodyWeight, titleWeight, metaWeight) +".txt");
					System.out.println("Write " + outputFile.getFileName().toString());
					
					Files.write(outputFile, output.getBytes());
				}
			}
		}
	}
	
	private final File runDir;
	private final String prefix;
	private final SharedTask task;
	private final Map<Integer, Map<String, Double>> topicToBodyScores;
	private final Map<Integer, Map<String, Double>> topicToTitleScores;
	private final Map<Integer, Map<String, Double>> topicToMetaScores;
	
	public CombineRankings(File runDir, String prefix, SharedTask task) {
		this.runDir = runDir;
		this.prefix = prefix;
		this.task = task;
		this.topicToBodyScores = loadScoresFromRunFile(runDir.toPath().resolve(prefix + "-body_lang.en^1-title_lang.en^0-meta_desc_lang.en^0"));
		this.topicToTitleScores = loadScoresFromRunFile(runDir.toPath().resolve(prefix + "-body_lang.en^0-title_lang.en^1-meta_desc_lang.en^0"));
		this.topicToMetaScores = loadScoresFromRunFile(runDir.toPath().resolve(prefix + "-body_lang.en^0-title_lang.en^0-meta_desc_lang.en^1"));
	}

	public String combine(float bodyWeight, float titleWeight, float metaWeight) {
		String ret = "";
		String runTag = runTag(prefix, bodyWeight, titleWeight, metaWeight);

		Map<String, List<RunLine>> topicToRunFiles = topicToRunFiles(task, bodyWeight, titleWeight, metaWeight);
		for (int qid : topics(topicToRunFiles)) {
			ScoredDocuments docs = adjustScoreTies(topicToRunFiles.get("" + qid));
			for (int i = 0; i < docs.documents.length; i++) {
				ret += "\n" + String.format(Locale.US, "%s Q0 %s %d %f %s", qid,
						docs.documents[i].getField(FIELD_ID).stringValue(), (i + 1), docs.scores[i], runTag);
			}
		}

		return ret.trim();
	}

	private Map<Integer, Map<String, Double>> loadScoresFromRunFile(Path runFile) {
		List<RunLine> lines = RunLine.parseRunlines(runFile);
		Map<Integer, Map<String, Double>> ret = new HashMap<>();
		
		for(RunLine line: lines) {
			if(!ret.containsKey(line.getTopic())) {
				ret.put(line.getTopic(), new LinkedHashMap<>());
			}

			ret.get(line.getTopic()).put(line.getDoucmentID(), score(line));
		}
		
		return ret;
	}
	
	private static String runTag(String prefix, float bodyWeight, float titleWeight, float metaWeight) {
		return prefix + "-body_lang.en^" + String.format("%.1f", bodyWeight) + "-title_lang.en^" 
				+ String.format("%.1f", titleWeight) + "-meta_desc_lang.en^" + String.format("%.1f", metaWeight);
	}

	private static List<Integer> topics(Map<String, List<RunLine>> topicToRunFiles) {
		return topicToRunFiles.keySet().stream().map(i -> Integer.parseInt(i)).sorted().collect(Collectors.toList());
	}

	private Map<String, List<RunLine>> topicToRunFiles(SharedTask task, float bodyWeight, float titleWeight, float metaWeight) {
		Map<String, List<RunLine>> ret = new LinkedHashMap<>();
		Set<QrelEqualWithoutScore> qrels = task.getQrelResourcesWithoutScore();
		
		for(QrelEqualWithoutScore qrel: qrels) {
			if(!ret.containsKey(qrel.getTopicNumber() + "")) {
				ret.put(qrel.getTopicNumber() + "", new ArrayList<>());
			}
			
			ret.get(qrel.getTopicNumber() + "").add(new RunLine(String.format(Locale.US, "%s Q0 %s %d %f %s",
					qrel.getTopicNumber(), qrel.getDocumentID(), 0, score(qrel.getTopicNumber(), qrel.getDocumentID(), bodyWeight, titleWeight, metaWeight), "tag")));
		}
		
		return ret;
	}

	private float score(int topicNumber, String documentID, float bodyWeight, float titleWeight, float metaWeight) {
		return (float) (topicToBodyScores.get(topicNumber).get(documentID) * bodyWeight
				+ topicToTitleScores.get(topicNumber).get(documentID) * titleWeight
				+ topicToMetaScores.get(topicNumber).get(documentID) * metaWeight);
	}

	private static ScoredDocuments adjustScoreTies(List<RunLine> runLines) {
		ScoredDocuments scoredDocs = docs(runLines);
		return new ScoreTiesAdjusterReranker().rerank(scoredDocs, null);
	}

	private static ScoredDocuments docs(List<RunLine> runLines) {
		runLines = new ArrayList<>(runLines);
		Collections.sort(runLines, (a, b) -> compare(b,a));
		ScoredDocuments scoredDocs = new ScoredDocuments();
		scoredDocs.documents = new Document[runLines.size()];
		scoredDocs.scores = new float[runLines.size()];

		for (int i = 0; i < runLines.size(); i++) {
			scoredDocs.documents[i] = doc(runLines.get(i));
			scoredDocs.scores[i] = score(runLines.get(i)).floatValue();
		}

		return scoredDocs;
	}
	
	private static int compare(RunLine a, RunLine b) {
		int ret = score(a).compareTo(score(b));
		
		if (ret != 0) {
			return ret;
		} else {
			return a.getDoucmentID().compareTo(b.getDoucmentID());
		}
	}

	private static Document doc(RunLine runLine) {
		Document ret = new Document();
		ret.add(new StringField(LuceneDocumentGenerator.FIELD_ID, runLine.getDoucmentID(), Store.YES));
		return ret;
	}

	private static Double score(RunLine runLine) {
		return Double.parseDouble(runLine.toString().split("[\\s\\t]+")[4]);
	}
}
