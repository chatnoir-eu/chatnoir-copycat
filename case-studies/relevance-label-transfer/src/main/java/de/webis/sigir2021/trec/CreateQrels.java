package de.webis.sigir2021.trec;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import de.webis.sigir2021.trec.LabelTransfer.RelevanceTransfer;
import de.webis.trec_ndd.trec_collections.QrelEqualWithoutScore;
import de.webis.trec_ndd.trec_collections.SharedTask;
import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.SneakyThrows;

public class CreateQrels {
	public static void main(String[] args) {
		Path baseDir = Paths.get("src/main/resources/artificial-qrels");
		
		writeDuplicateFreeQrels(TrecSharedTask.WEB_2009, baseDir.resolve("qrels.inofficial.duplicate-free.web.1-50.txt"));
		writeDuplicateFreeQrels(TrecSharedTask.WEB_2010, baseDir.resolve("qrels.inofficial.duplicate-free.web.51-100.txt"));
		writeDuplicateFreeQrels(TrecSharedTask.WEB_2011, baseDir.resolve("qrels.inofficial.duplicate-free.web.101-150.txt"));
		writeDuplicateFreeQrels(TrecSharedTask.WEB_2012, baseDir.resolve("qrels.inofficial.duplicate-free.web.151-200.txt"));
		writeDuplicateFreeQrels(TrecSharedTask.WEB_2013, baseDir.resolve("qrels.inofficial.duplicate-free.web.201-250.txt"));
		writeDuplicateFreeQrels(TrecSharedTask.WEB_2014, baseDir.resolve("qrels.inofficial.duplicate-free.web.251-300.txt"));
		
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cw12.1-50.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cw12.51-100.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cw12.101-150.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cw12.151-200.jsonl"));
		
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cw12wb12.1-50.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cw12wb12.51-100.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cw12wb12.101-150.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cw12wb12.151-200.jsonl"));
		
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cc15.1-50.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cc15.51-100.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cc15.101-150.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cc15.151-200.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cc15.201-250.jsonl"));
		jsonlToQrels(baseDir.resolve("qrels.inofficial.duplicate-free.transferred-to-cc15.251-300.jsonl"));
	}
	
	@SneakyThrows
	private static void jsonlToQrels(Path jsonlFile) {
		Path targetFile = jsonlFile.getParent().resolve(jsonlFile.getFileName().toString().replaceAll("\\.jsonl", ".txt"));
		List<QrelEqualWithoutScore> lines = Files.readAllLines(jsonlFile).stream().map(i -> jsonToQrel(i)).collect(Collectors.toList());
		//FIXME: Do sanity-checks
		String ret = lines.stream().filter(i -> i != null).map(i-> i.toString()).collect(Collectors.joining("\n"));
		Files.write(targetFile, ret.getBytes());
	}

	private static QrelEqualWithoutScore jsonToQrel(String json) {
		RelevanceTransfer rt = RelevanceTransfer.fromString(json);
		if(rt.getTargetDocs() == null || rt.getTargetDocs().isEmpty()) {
			return null;
		}
		
		QrelEqualWithoutScore ret = new QrelEqualWithoutScore();
		ret.setDocumentID(rt.getTargetDocs().get(0).getTargetId());
		ret.setScore(rt.getRelevanceLabel());
		ret.setTopicNumber(rt.getTopicNumber());
		
		return ret;
	}

	@SneakyThrows
	private static void writeDuplicateFreeQrels(SharedTask task, Path path) {
		String qrels = duplicateFreeQrels(task).stream().collect(Collectors.joining("\n"));
		Files.write(path, qrels.getBytes());
	}
	
	private static List<String> duplicateFreeQrels(SharedTask task) {
		List<String> ret = new ArrayList<>();
		
		for(QrelEqualWithoutScore qrel: task.getQrelResourcesWithoutScore()) {
			if(!CanonicalDocuments.isDuplicate(qrel.getDocumentID())) {
				System.out.println(qrel);
				ret.add(qrel.toString());
			}
		}
		
		return ret;
	}
}
