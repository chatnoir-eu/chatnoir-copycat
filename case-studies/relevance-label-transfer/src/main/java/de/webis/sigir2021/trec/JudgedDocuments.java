package de.webis.sigir2021.trec;

import java.time.temporal.ChronoUnit;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import org.apache.http.HttpHost;
import org.elasticsearch.action.get.GetRequest;
import org.elasticsearch.action.get.GetResponse;
import org.elasticsearch.client.RequestOptions;
import org.elasticsearch.client.RestClient;
import org.elasticsearch.client.RestHighLevelClient;

import de.webis.WebisUUID;
import de.webis.trec_ndd.trec_collections.SharedTask;
import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;
import net.sourceforge.argparse4j.inf.Namespace;

public class JudgedDocuments {

	public static final RetryPolicy<String> RETRY_POLICY = new RetryPolicy<String>()
			.handle(Exception.class)
			.withBackoff(10, 240, ChronoUnit.SECONDS)
			.withMaxRetries(5);
	
	public static String urlOfDocument(String docId) {
		return Failsafe.with(RETRY_POLICY).get(() -> urlOfDocumentThrowsException(docId));
	}
	
	@SneakyThrows
	public static String urlOfDocumentThrowsException(String docId) {
		try(RestHighLevelClient client = client()) {
			GetResponse response = client.get(new GetRequest(index(docId), type(docId), chatNoirId(docId)), RequestOptions.DEFAULT);
			
			if(response.isExists()) {
				return response.getSource().get("warc_target_uri").toString();
			} else {
				return null;
			}
		}
	}
	
	public static String chatNoirId(String documentId) {
		return chatNoirId(longChatNoirId(documentId), documentId).toString();
	}
	
	public static UUID chatNoirId(String prefix, String documentId) {
		return new WebisUUID(prefix).generateUUID(documentId);
	}
	
	private static String longChatNoirId(String documentId) {
		if (documentId.startsWith("clueweb09")) {
			return "clueweb09";
		} else if (documentId.startsWith("clueweb12")) {
			return "clueweb12";
		}

		throw new RuntimeException("ID '" + documentId + "' is not supported.");
	}
	
	public static String index(String documentId) {
		if (documentId.startsWith("clueweb09")) {
			return "webis_warc_clueweb09_003";
		} else if (documentId.startsWith("clueweb12")) {
			return "webis_warc_clueweb12_011";
		}

		throw new RuntimeException("ID '" + documentId + "' is not supported.");
	}

	public static String type(String documentId) {
		return "warcrecord";
	}
	
	private static RestHighLevelClient client() {
		return new RestHighLevelClient(RestClient.builder(new HttpHost("betaweb023", 9200, "http")));
	}

	public static List<String> judgedDocuments(Namespace parsedArgs) {
		SharedTask task = task(parsedArgs);
		Integer topic = parsedArgs.getInt("topic");
		Set<String> ret = new HashSet<>();
		
		ret.addAll(task.documentJudgments().getRelevantDocuments("" + topic));
		ret.addAll(task.documentJudgments().getIrrelevantDocuments("" + topic));
		
		return ret.stream().sorted().collect(Collectors.toList());
	}
	
	private static SharedTask task(Namespace parsedArgs) {
		return TrecSharedTask.valueOf(parsedArgs.getString("task"));
	}
}
