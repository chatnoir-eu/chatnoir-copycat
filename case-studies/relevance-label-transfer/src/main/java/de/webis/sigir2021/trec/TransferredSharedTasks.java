package de.webis.sigir2021.trec;

import java.util.List;
import java.util.Map;

import de.webis.trec_ndd.trec_collections.SharedTask;
import de.webis.trec_ndd.trec_eval.EvaluationMeasure;

public class TransferredSharedTasks {
	public static final SharedTask WEB_2009_DUPLICATE_FREE = task("/artificial-qrels/qrels.inofficial.duplicate-free.web.1-50.txt", 2009),
								  WEB_2010_DUPLICATE_FREE = task("/artificial-qrels/qrels.inofficial.duplicate-free.web.51-100.txt", 2010),
								  WEB_2011_DUPLICATE_FREE = task("/artificial-qrels/qrels.inofficial.duplicate-free.web.101-150.txt", 2011),
								  WEB_2012_DUPLICATE_FREE = task("/artificial-qrels/qrels.inofficial.duplicate-free.web.151-200.txt", 2012);
	
	public static final SharedTask TRANSFER_FROM_WEB_2009_TO_CW_2012 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cw12.1-50.txt", 2009),
								  TRANSFER_FROM_WEB_2010_TO_CW_2012 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cw12.51-100.txt", 2010),
								  TRANSFER_FROM_WEB_2011_TO_CW_2012 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cw12.101-150.txt", 2011),
								  TRANSFER_FROM_WEB_2012_TO_CW_2012 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cw12.151-200.txt", 2012);
	

	public static final SharedTask TRANSFER_FROM_WEB_2009_TO_CW_WB_2012 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cw12wb12.1-50.txt", 2009),
								  TRANSFER_FROM_WEB_2010_TO_CW_WB_2012 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cw12wb12.51-100.txt", 2010),
								  TRANSFER_FROM_WEB_2011_TO_CW_WB_2012 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cw12wb12.101-150.txt", 2011),
								  TRANSFER_FROM_WEB_2012_TO_CW_WB_2012 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cw12wb12.151-200.txt", 2012);

	public static final SharedTask WEB_2013_DUPLICATE_FREE = task("/artificial-qrels/qrels.inofficial.duplicate-free.web.201-250.txt", 2013),
								  WEB_2014_DUPLICATE_FREE = task("/artificial-qrels/qrels.inofficial.duplicate-free.web.251-300.txt", 2014);
	

	public static final SharedTask TRANSFER_FROM_WEB_2009_TO_CC_2015 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cc15.1-50.txt", 2009),
								  TRANSFER_FROM_WEB_2010_TO_CC_2015 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cc15.51-100.txt", 2010),
								  TRANSFER_FROM_WEB_2011_TO_CC_2015 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cc15.101-150.txt", 2011),
								  TRANSFER_FROM_WEB_2012_TO_CC_2015 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cc15.151-200.txt", 2012),
								  TRANSFER_FROM_WEB_2013_TO_CC_2015 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cc15.201-250.txt", 2013),
								  TRANSFER_FROM_WEB_2014_TO_CC_2015 = task("/artificial-qrels/qrels.inofficial.duplicate-free.transferred-to-cc15.251-300.txt", 2014);
	
	private static SharedTask task(String qrelResource, int year) {
		return new SharedTask() {
			@Override
			public String getQrelResource() {
				return qrelResource;
			}

			@Override
			public List<EvaluationMeasure> getOfficialEvaluationMeasures() {
				throw new RuntimeException("TBD");
			}

			@Override
			public List<EvaluationMeasure> getInofficialEvaluationMeasures() {
				throw new RuntimeException("TBD");
			}

			@Override
			public List<String> runFiles() {
				throw new RuntimeException("TBD");
			}

			@Override
			public String name() {
				throw new RuntimeException("TBD");
			}

			@Override
			public Map<String, Map<String, String>> topicNumberToTopic() {
				return TrecSharedTask.valueOf("WEB_" + year).topicNumberToTopic();
			}
		};
	}
}
