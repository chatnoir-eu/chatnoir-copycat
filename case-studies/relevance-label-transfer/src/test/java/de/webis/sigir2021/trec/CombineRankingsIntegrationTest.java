package de.webis.sigir2021.trec;

import java.io.File;
import java.util.List;
import java.util.Map;

import org.approvaltests.Approvals;
import org.junit.Test;

import de.webis.sigir2021.trec.CombineRankings;
import de.webis.trec_ndd.trec_collections.SharedTask;
import de.webis.trec_ndd.trec_eval.EvaluationMeasure;

public class CombineRankingsIntegrationTest {
	@Test
	public void approveAllFieldsSetToZero() {
		File runDir = new File("src/test/resources/sample-run-dir");
		String prefix = "webis_warc_clueweb12_011";
		SharedTask task = testSharedTask();
		
		String combinedRanking = new CombineRankings(runDir, prefix, task).combine(0, 0, 0);
		
		Approvals.verify(combinedRanking);
	}
	
	@Test
	public void approveMetaWeight1() {
		File runDir = new File("src/test/resources/sample-run-dir");
		String prefix = "webis_warc_clueweb12_011";
		SharedTask task = testSharedTask();
		
		String combinedRanking = new CombineRankings(runDir, prefix, task).combine(0, 0, 1);
		
		Approvals.verify(combinedRanking);
	}
	
	@Test
	public void approveTitleWeight1() {
		File runDir = new File("src/test/resources/sample-run-dir");
		String prefix = "webis_warc_clueweb12_011";
		SharedTask task = testSharedTask();
		
		String combinedRanking = new CombineRankings(runDir, prefix, task).combine(0, 1, 0);
		
		Approvals.verify(combinedRanking);
	}
	
	@Test
	public void approveBodyWeight1() {
		File runDir = new File("src/test/resources/sample-run-dir");
		String prefix = "webis_warc_clueweb12_011";
		SharedTask task = testSharedTask();
		
		String combinedRanking = new CombineRankings(runDir, prefix, task).combine(1, 0, 0);
		
		Approvals.verify(combinedRanking);
	}
	
	@Test
	public void approveAllFields1() {
		File runDir = new File("src/test/resources/sample-run-dir");
		String prefix = "webis_warc_clueweb12_011";
		SharedTask task = testSharedTask();
		
		String combinedRanking = new CombineRankings(runDir, prefix, task).combine(1, 1, 1);
		
		Approvals.verify(combinedRanking);
	}

	private static SharedTask testSharedTask() {
		return new SharedTask() {
			@Override
			public String getQrelResource() {
				return "/sample-run-dir/qrels-test.txt";
			}

			@Override
			public List<EvaluationMeasure> getOfficialEvaluationMeasures() {
				throw new RuntimeException("TBD");
			}

			@Override
			public List<EvaluationMeasure> getInofficialEvaluationMeasures() {
				throw new RuntimeException("TBD");
			}

			@Override
			public List<String> runFiles() {
				throw new RuntimeException("TBD");
			}

			@Override
			public String name() {
				throw new RuntimeException("TBD");
			}

			@Override
			public Map<String, Map<String, String>> topicNumberToTopic() {
				return null;
			}
		};
	}
}
