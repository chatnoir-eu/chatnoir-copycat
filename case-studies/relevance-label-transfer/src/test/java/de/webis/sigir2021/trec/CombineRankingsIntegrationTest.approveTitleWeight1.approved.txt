23 Q0 clueweb12-1310wb-04-20329 1 0.000000 webis_warc_clueweb12_011-body_lang.en^0.0-title_lang.en^1.0-meta_desc_lang.en^0.0
23 Q0 clueweb12-1308wb-87-17434 2 -0.000001 webis_warc_clueweb12_011-body_lang.en^0.0-title_lang.en^1.0-meta_desc_lang.en^0.0
23 Q0 clueweb12-0000wb-70-08228 3 -0.000002 webis_warc_clueweb12_011-body_lang.en^0.0-title_lang.en^1.0-meta_desc_lang.en^0.0
131 Q0 clueweb12-0108wb-19-02294 1 3.000000 webis_warc_clueweb12_011-body_lang.en^0.0-title_lang.en^1.0-meta_desc_lang.en^0.0
131 Q0 clueweb12-0403wb-55-25772 2 0.000000 webis_warc_clueweb12_011-body_lang.en^0.0-title_lang.en^1.0-meta_desc_lang.en^0.0
131 Q0 clueweb12-0110wb-61-03700 3 -0.000001 webis_warc_clueweb12_011-body_lang.en^0.0-title_lang.en^1.0-meta_desc_lang.en^0.0